aiida/storage/psql_dos/migrator.py

```diff
@@ -148,47 +146,144 @@ def validate_storage(self) -> None:
+
+    def initialise(self, reset: bool = False) -> None:
+        """Initialise the storage backend.
+
+        This is typically used once when a new storage backed is created. If this method returns without exceptions the
+        storage backend is ready for use. If the backend already seems initialised, this method is a no-op.
+
+        :param reset: If ``true``, destroy the backend if it already exists including all of its data before recreating
+            and initialising it. This is useful for example for test profiles that need to be reset before or after
+            tests having run.
+        """
+        if reset:
+            self.reset_repository()
+            self.reset_database()
+
+        if not self.is_initialised:
+            self.initialise_repository()
+            self.initialise_database()
+
+        self.migrate()
+
+    @property
+    def is_initialised(self) -> bool:
+        """Return whether the storage is initialised.
+
+        This is the case if both the database and the repository are initialised.
+
+        :returns: ``True`` if the storage is initialised, ``False`` otherwise.
+        """
+        return self.is_repository_initialised and self.is_database_initialised
+
+    @property
+    def is_repository_initialised(self) -> bool:
+        """Return whether the repository is initialised.
+
+        :returns: ``True`` if the repository is initialised, ``False`` otherwise.
+        """
+        return self.get_container().is_initialised
+
+    @property
+    def is_database_initialised(self) -> bool:
+        """Return whether the database is initialised.
+
+        This is the case if it contains the table that holds the schema version for alembic or Django.
+
+        :returns: ``True`` if the database is initialised, ``False`` otherwise.
+        """
+        with self._connection_context() as connection:
+            return (
+                inspect(connection).has_table(self.alembic_version_tbl_name) or
+                inspect(connection).has_table(self.django_version_table.name)
+            )
+
+    def reset_repository(self) -> None:
+        """Reset the repository by deleting all of its contents.
+
+        This will also destroy the configuration and so in order to use it again, it will have to be reinitialised.
+        """
+        import shutil
+        try:
+            shutil.rmtree(self.get_container().get_folder())
+        except FileNotFoundError:
+            pass
+
+    def reset_database(self) -> None:
+        """Reset the database by deleting all content from all tables.
+
+        This will also destroy the settings table and so in order to use it again, it will have to be reinitialised.
+        """
+        with self._connection_context() as connection:
+            if inspect(connection).has_table(self.alembic_version_tbl_name):
+                for table_name in (
+                    'db_dbgroup_dbnodes', 'db_dbgroup', 'db_dblink', 'db_dbnode', 'db_dblog', 'db_dbauthinfo',
```

Comment:
just checking whether this list of tables can either be deduced from somewhere (or, if not, whether it should be stored in a reusable constant)

Reply:
Not sure to be honest. There is one similar list in the `PsqlDosBackend._clear` method minus the `db_dbsetting` table. I tried to change as little as possible and just get the minimum that is required.

Reply:
yeh it should likely be deduced by reflection

Reply:
Done. Also changed the backend to call the implementation of the migrator, since that was also hard-coding the tables in the `_clear` method.

Reply:
It seems that the new introspective method may be considerably slower. Not sure if the tests being slower is a fluke or due to this change (test duration varies quite a bit). Also seems a problem with too many connections being opened.

Reply:
Can you reproduce the change in test duration locally?

Reply:
I have isolated the changes that I think are responsible. Just pushed the commit without those changes to see if the test times return to normal. Then I will add the isolated changes in a separate commit and run again. That should give us a good sense whether the change in time is real. Also easier to review the changes and spot a mistake in connection handling.

Reply:
Oh yeh, I remember there being issues with reflection and opening/closing sessions

Reply:
Ok, without the table delete changes, test times are normal (~13-14 mins): https://github.com/sphuber/aiida-core/actions/runs/3479079616/jobs/5817199424